src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,364 @@
+        }
+
+        // This is a redeclaration of the new type in the .NET 8 TFM.
+        private enum UnsafeAccessorKind
+        {
+            Constructor,
+            Method,
+            StaticMethod,
+            Field,
+            StaticField
+        };
+
+        private static bool TryParseUnsafeAccessorAttribute(MethodDesc method, CustomAttributeValue<TypeDesc> decodedValue, out UnsafeAccessorKind kind, out string name)
+        {
+            kind = default;
+            name = default;
+
+            var context = method.Context;
+
+            // Get the kind of accessor
+            if (decodedValue.FixedArguments.Length != 1
+                || decodedValue.FixedArguments[0].Type.UnderlyingType != context.GetWellKnownType(WellKnownType.Int32))
+            {
+                return false;
+            }
+
+            kind = (UnsafeAccessorKind)decodedValue.FixedArguments[0].Value!;
+
+            // Check the name of the target to access. This is the name we
+            // use to look up the intended token in metadata.
+            string nameMaybe = null;
+            foreach (var argument in decodedValue.NamedArguments)
+            {
+                if (argument.Name == "Name")
+                {
+                    nameMaybe = (string)argument.Value;
+                }
+            }
+
+            // If the Name isn't defined, then use the name of the method.
+            if (nameMaybe != null)
+            {
+                name = nameMaybe;
+            }
+            else
+            {
+                // The Constructor case has an implied value provided by
+                // the runtime. We are going to enforce this during consumption
+                // so we avoid the setting of the value. We validate the name
+                // as empty at the use site.
+                if (kind is not UnsafeAccessorKind.Constructor)
+                {
+                    name = nameMaybe;
+                }
+            }
+
+            return true;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct GenerationContext
+        {
+            public UnsafeAccessorKind Kind;
+            public EcmaMethod Declaration;
+            public TypeDesc TargetType;
+            public bool IsTargetStatic;
+            public MethodDesc TargetMethod;
+            public FieldDesc TargetField;
+        }
+
+        private static bool TrySetTargetMethod(ref GenerationContext context, string name)
+        {
+            TypeDesc targetType = context.TargetType.IsByRef
+                ? ((ParameterizedType)context.TargetType).ParameterType
+                : context.TargetType;
+
+            // Due to how some types degrade, unsafe access on these
+            // target types is blocked, even if the lookup succeeds.
+            // For example, pointers can become lookups on typeof(uint)'s MethodTable.
+            if (targetType.IsParameterizedType)
```

Comment:
```suggestion
            if ((targetType.IsParameterizedType && targetType.IsArray) || targetType.IsFunctionPointerType)
```
To match the unmanaged impl